core/src/main/java/org/apache/accumulo/core/clientImpl/ThriftTransportPool.java

```diff
@@ -67,7 +70,8 @@ public CachedConnection reserveAny() {
     }
   }
 
-  private Map<ThriftTransportKey,CachedConnections> cache = new HashMap<>();
+  private ConcurrentHashMap<ThriftTransportKey,Lock> locks = new ConcurrentHashMap<>();
```

Comment:
I think these need to be volatile. Also a single volatile pointer to `locks` and `cache` may help with the fact that shutdown can now run concurrently with connection reservation and allocation. 

Maybe something like 

```java
   private static class LocksAndCache {
      ConcurrentHashMap<ThriftTransportKey,Lock> locks = new ConcurrentHashMap<>();
      ConcurrentHashMap<ThriftTransportKey,CachedConnections> cache = new ConcurrentHashMap<>();
   }

  private volatile LocksAndCache locksAndCache = new LocksAndCahce();
```

I'll explain in other comments why I think this may be useful.

Reply:
I created something similar to this, but called `ConnectionPool`. I also added on additional methods both for `ConnectionPool` and `CachedConnections` in an attempt to make the code more concise overall.

Reply:
I like the changes you made.